ophyd/v2/core.py

```diff
@@ -403,86 +408,151 @@ def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
+        self._severity = 0
+        self._reading = self.converter.reading(
+            self._initial_value, time.monotonic(), self._severity
+        )
 
-    async def connect(self):
-        pass
+        await self.put(self._initial_value)
 
     async def put(self, value: Optional[T], wait=True, timeout=None):
         if value is None:
-            self.set_value(self._initial_value)
+            write_value = self._initial_value
         else:
-            self.set_value(value)
+            write_value = self.converter.write_value(value)
+
+        self._value = write_value
+        self._timestamp = time.monotonic()
+        self._reading = await self.get_reading()
+        if self.callback:
+            self.callback(self._reading, self._value)
+
         if wait:
             await asyncio.wait_for(self.put_proceeds.wait(), timeout)
 
     async def get_descriptor(self) -> Descriptor:
-        try:
-            dtype = primitive_dtypes[type(self._value)]
-            shape = []
-        except KeyError:
-            if isinstance(self._value, Sequence):
-                dtype = "array"
-                shape = [len(self._value)]
-            elif isinstance(self._value, Enum):
-                dtype = "string"
-                shape = []
-                choices = [e.value for e in type(self._value)]
-                # Ignore type until https://github.com/bluesky/event-model/issues/235#issuecomment-1497353265
-                return dict(source=self.source, dtype=dtype, shape=shape, choices=choices)  # type: ignore
-            else:
-                assert False, f"Can't get dtype for {type(self._value)}"
-        return dict(source=self.source, dtype=dtype, shape=shape)
-
-    @property
-    def _reading(self) -> Reading:
-        return dict(value=self._value, timestamp=self._timestamp)
+        return self.converter.descriptor(self.source, self._value)
 
     async def get_reading(self) -> Reading:
-        return self._reading
+        return self.converter.reading(self._value, self._timestamp, self._severity)
 
     async def get_value(self) -> T:
-        return self._value
+        return self.converter.value(self._value)
 
     def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
         if callback:
             assert not self.callback, "Cannot set a callback when one is already set"
             callback(self._reading, self._value)
         self.callback = callback
 
-    def set_value(self, value: T) -> None:
-        """Set the simulated value, and set timestamp to now"""
-        self._value = value
-        self._timestamp = time.monotonic()
-        if self.callback:
-            self.callback(self._reading, self._value)
-
 
-def set_sim_value(signal: Signal[T], value: T):
+async def set_sim_value(signal: Signal[T], value: T):
     """Set the value of a signal that is in sim mode"""
-    _sim_backends[signal].set_value(value)
+    await _sim_backends[signal].put(value)
```

Comment:
Please can you turn this back to a regular function rather than an async one, it's important that we can set the simulated value of a signal without yielding to asyncio so we can do some of the more tricky sequencing of tests

Reply:
All the tests are passing; what kind of sequencing are we talking about?

I think it makes the SimSignalBackend messier and less like it's non-simulated counterparts.

If you'd like I can write a test that means I have to use a non-async function. Although there are no such existing tests and you'd have to describe a situation which would warrant this.